database/src/lib.rs

```diff
@@ -72,6 +72,159 @@
+                RetryDecision::RetryNextNode(None)
+            }
+
+            // Unavailable - the current node believes that not enough nodes
+            // are alive to satisfy specified consistency requirements.
+            // Maybe this node has network problems - try a different one.
+            // Perform at most one retry - it's unlikely that two nodes
+            // have network problems at the same time
+            QueryError::DbError(scylla::transport::errors::DbError::Unavailable { .. }, _) => {
+                if self.is_retry() {
+                    RetryDecision::RetryNextNode(None)
+                } else {
+                    RetryDecision::DontRetry
+                }
+            }
+
+            // ReadTimeout - coordinator didn't receive enough replies in time.
+            // Retry at most once and only if there were actually enough replies
+            // to satisfy consistency but they were all just checksums (data_present == false).
+            // This happens when the coordinator picked replicas that were overloaded/dying.
+            // Retried request should have some useful response because the node will detect
+            // that these replicas are dead.
+            QueryError::DbError(
+                scylla::transport::errors::DbError::ReadTimeout {
+                    received,
+                    required,
+                    data_present,
+                    ..
+                },
+                _,
+            ) => {
+                if self.is_retry() && received >= required && !*data_present {
+                    RetryDecision::RetrySameNode(None)
+                } else {
+                    RetryDecision::DontRetry
+                }
+            }
+
+            // Write timeout - coordinator didn't receive enough replies in time.
+            // Retry at most once and only for BatchLog write.
+            // Coordinator probably didn't detect the nodes as dead.
+            // By the time we retry they should be detected as dead.
+            QueryError::DbError(scylla::transport::errors::DbError::WriteTimeout { .. }, _) => {
+                if self.is_retry() {
+                    RetryDecision::RetryNextNode(None)
+                } else {
+                    RetryDecision::DontRetry
+                }
+            }
+            // The node is still bootstrapping it can't execute the query, we should try another one
+            QueryError::DbError(scylla::transport::errors::DbError::IsBootstrapping, _) => {
+                RetryDecision::RetryNextNode(None)
+            }
+            // Connection to the contacted node is overloaded, try another one
+            QueryError::UnableToAllocStreamId => RetryDecision::RetryNextNode(None),
+            // In all other cases propagate the error to the user
+            _ => RetryDecision::DontRetry,
+        }
+    }
+
+    /// Reset before using for a new query
+    fn reset(&mut self) {
```

Comment:
`reset_max_retry_counter`

Reply:
This is method of `scylla::retry_policy::RetrySession` trait. We can't rename it.